src/test/ui/cast/casts-issue-46365.rs

```diff
@@ -3,5 +3,5 @@ struct Lorem {
 }
 
 fn main() {
-    let _foo: *mut Lorem = core::ptr::null_mut(); // no error here
+    let _foo: *mut Lorem = 0 as *mut _; // no error here
```

Comment:
Curious: If you made a field with a known non-DST type after `ipsum` in `Lorem`, would it not get tripped up? It does seem like since `ipsum` is the final field, it can't derive the metadata one way or another, but otherwise it perhaps could.

Reply:
With strict provenance (which I believe was added after this comment) we definitely want to keep `null_mut` instead of zero casts.

Reply:
Would `null_mut` itself not be implemented with a cast under strict provenance?

Reply:
null_mut defers to invalid_mut which in turn is just a cast yeah. The ideal is for everything to funnel into methods with clearer intent/semantics so that automated tooling better understands what you're trying to do and can help catch bugs (i.e. maybe invalid_mut becomes an intrinsic so the compiler can mark some metadata for other tools? 

Also more stuff using the "right" APIs for what they want to do makes it easier for cursed platforms to do hacks. Like if someone *really* wanted to force Rust onto a null!=0 platform, everything going through null/null_mut/is_null makes that a lot more plausible of an officially-unsupported-but-functional-hack.

Reply:
I’ve changed this file to use `NonNull::dangling` to preserve the spirit of the test while avoiding the unrelated error that this PR adds with `null_mut` at a line commented with `// no error here`

Reply:
FWIW the original bug (https://github.com/rust-lang/rust/issues/46365) was about `as` casts so I highly doubt that this is still testing what it was intended to test (which is a problem introduced by https://github.com/rust-lang/rust/pull/61864 doing blind search-and-replace).

Reply:
Opened https://github.com/rust-lang/rust/pull/97765 to restore test.